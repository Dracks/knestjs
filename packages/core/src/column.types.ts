
export type ColumnType = 'int' | 'float' | 'double' | 'char' | 'varchar' | 'text' | 'date' | 'datetime' | 'time'


export interface IntColumnConfig  {
    type: 'int'
    autoincrement?: boolean
}

export interface FloatColumnConfig  {
    type: 'float',
    precision: number,
    scale: number
}

export interface CharColumnConfig {
    type: 'char' | 'varchar'
    length: number
}

export interface DateTimeColumnConfig {
    type: 'datetime' | 'date' | 'time'
}
