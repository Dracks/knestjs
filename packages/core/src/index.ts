export * from './knest.module'
export * from './migrations/migrations.service'
export * from './decorators/table.decorator'
export * from './decorators/column.decorator'
export {KNEX_INSTANCE} from './constants'
