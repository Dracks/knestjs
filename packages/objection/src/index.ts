export * from './knestjs-objection.module';
export * from './types';
