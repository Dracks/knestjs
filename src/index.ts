export * from './knest.module'
export * from './services/migrations.service'
export * from './decorators/table.decorator'
export * from './decorators/column.decorator'
export {KNEX_INSTANCE} from './constants'
