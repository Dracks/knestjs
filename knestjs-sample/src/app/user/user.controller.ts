import { Controller, Get, Inject, OnModuleInit } from "@nestjs/common";
import { Knex } from "knex";
import { KNEX_INSTANCE } from '@knestjs/core'
import { UserModel } from "./user.model";

@Controller('users')
export class UserController{
    constructor(@Inject(KNEX_INSTANCE) readonly knex: Knex){}

    @Get('list')
    listUsers(){
        return this.knex.select<UserModel>().from('user')
    }
}
