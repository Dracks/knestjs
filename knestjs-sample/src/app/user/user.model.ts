import { Column, Table } from "@knestjs/core";

@Table({
    name: 'user'
})
export class UserModel {
    @Column({
        type: 'int',
        autoincrement: true,
        nullable: false
    })
    id!: number

    @Column({
        type: 'varchar',
        length: 255,
        nullable: false
    })
    user!: string;
}
