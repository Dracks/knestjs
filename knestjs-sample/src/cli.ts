import { NestFactory } from '@nestjs/core';
import { CommandModule, CommandService } from 'nestjs-command';
import { CliModule } from './cli/cli.module';

async function bootstrap () {
    const app = await NestFactory.createApplicationContext(CliModule);

    try {
        await app
          .select(CommandModule)
          .get(CommandService)
          .exec();
    } catch (error) {
        console.error(error);
    } finally{
        await app.close()
    }
}

bootstrap();
