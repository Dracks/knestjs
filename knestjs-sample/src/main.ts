import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';
import { KNEX_INSTANCE, MigrationsService} from '@knestjs/core'
import { UserModel } from './app/user/user.model';
import { Knex } from 'knex';


async function bootstrap() {
  const app = await NestFactory.create(AppModule);

  // migrate the database to the lates version
  await app.get(MigrationsService).latest()

  // Add some data into the database
  const knex : Knex = app.get(KNEX_INSTANCE)
  await knex<UserModel>('user').insert<UserModel>({
        user: 'John'
  })
  await knex<UserModel>('user').insert<UserModel>({
        user: 'Marta'
  })

  await app.listen(3000);
  console.log("App started!")
}
bootstrap();
