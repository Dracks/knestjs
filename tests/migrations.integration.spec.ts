import {Test, TestingModule} from '@nestjs/testing'
import {INestApplication} from '@nestjs/common'
import * as path from 'path'
import {promises as fs} from 'fs'

import {KnestModule, MigrationsService} from '../src'
import { UserModel } from './models/user.model'
import {oldSnapshot1} from './mocks/previous-version'

import 'reflect-metadata'

const migrationsFolder = path.join(__dirname, 'migrations');
const snapshotName = 'db-status.snapshot'

const getSnapshot = async ()=>{
    const contents = await fs.readFile(path.join(migrationsFolder,snapshotName))
    return JSON.parse(contents.toString())
}

describe('Integration tests of migrations', ()=>{
    let app!: INestApplication;

    describe('makemigrations', ()=>{
        beforeEach(async ()=>{
            const moduleFixture: TestingModule = await Test.createTestingModule({
                imports: [KnestModule.forRoot({
                    db: {
                        client: 'sqlite3',
                        connection: {
                            filename: ':memory:'
                        },
                        useNullAsDefault: true,
                    },
                    migrations: {
                        folder: migrationsFolder,
                        snapshotName,
                    }
                }),
                KnestModule.forFeature([UserModel])
            ],
            }).compile();

            app = moduleFixture.createNestApplication();
        })

        afterEach(async ()=>{
            await fs.rm(migrationsFolder, {recursive: true})
        })

        it('New database from 0', async ()=>{
            await app.init()

            expect(app).toBeTruthy()

            await app.get(MigrationsService).makeMigrations()

            expect(await getSnapshot()).toMatchSnapshot()
        })

        it('From an existing database', async ()=>{
          await fs.mkdir(migrationsFolder)
          await fs.writeFile(path.join(migrationsFolder, snapshotName), JSON.stringify(oldSnapshot1))
          await app.init()

          expect(app).toBeTruthy()

          await app.get(MigrationsService).makeMigrations()

          expect(await getSnapshot()).toMatchSnapshot()
        })
    })

})
