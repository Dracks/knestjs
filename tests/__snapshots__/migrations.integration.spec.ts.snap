// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests of migrations makemigrations From an existing database 1`] = `
{
  "db": {
    "UserModel": {
      "className": "UserModel",
      "columns": [
        {
          "autoincrement": true,
          "name": "id",
          "nullable": false,
          "primaryKey": true,
          "property": "id",
          "type": "int",
        },
        {
          "length": 255,
          "name": "user",
          "nullable": false,
          "property": "user",
          "type": "varchar",
        },
        {
          "length": 51,
          "name": "password",
          "nullable": true,
          "property": "password",
          "type": "char",
        },
        {
          "name": "age",
          "nullable": true,
          "property": "age",
          "type": "int",
        },
        {
          "name": "created",
          "nullable": false,
          "property": "created",
          "type": "datetime",
        },
      ],
      "indexes": [
        {
          "name": "user_idx",
          "properties": [
            "user",
          ],
        },
      ],
      "name": "UserModel",
    },
  },
  "knestVersion": "",
  "version": 2,
}
`;

exports[`Integration tests of migrations makemigrations From an existing database 2`] = `
"
exports.up = (knex) => knex.schema.alterTable("UserModel", (table) => {
    table.integer("age").nullable();
    table.dropColumn("to-remove");
    table.dropIndex(["something-to-remove"], "str_idx");
});

exports.down = (knex) => knex.schema.alterTable("UserModel", (table) => {
    table.string("to-remove", 10).nullable();
    table.dropColumn("age");
    table.index(["something-to-remove"], "str_idx");
});
"
`;

exports[`Integration tests of migrations makemigrations Multiple tables from 0 1`] = `
{
  "db": {
    "GroupModel": {
      "className": "GroupModel",
      "columns": [
        {
          "autoincrement": true,
          "name": "id",
          "nullable": false,
          "primaryKey": true,
          "property": "id",
          "type": "int",
        },
        {
          "length": 255,
          "name": "name",
          "nullable": false,
          "property": "name",
          "type": "varchar",
        },
        {
          "length": 51,
          "name": "privileges",
          "nullable": false,
          "property": "privileges",
          "type": "char",
        },
        {
          "name": "created",
          "nullable": false,
          "property": "created",
          "type": "datetime",
        },
      ],
      "indexes": [
        {
          "properties": [
            "privileges",
          ],
        },
      ],
      "name": "GroupModel",
    },
    "UserModel": {
      "className": "UserModel",
      "columns": [
        {
          "autoincrement": true,
          "name": "id",
          "nullable": false,
          "primaryKey": true,
          "property": "id",
          "type": "int",
        },
        {
          "length": 255,
          "name": "user",
          "nullable": false,
          "property": "user",
          "type": "varchar",
        },
        {
          "length": 51,
          "name": "password",
          "nullable": true,
          "property": "password",
          "type": "char",
        },
        {
          "name": "age",
          "nullable": true,
          "property": "age",
          "type": "int",
        },
        {
          "name": "created",
          "nullable": false,
          "property": "created",
          "type": "datetime",
        },
      ],
      "indexes": [
        {
          "name": "user_idx",
          "properties": [
            "user",
          ],
        },
      ],
      "name": "UserModel",
    },
  },
  "knestVersion": "",
  "version": 1,
}
`;

exports[`Integration tests of migrations makemigrations Multiple tables from 0 2`] = `
"
exports.up = (knex) => knex.schema.createTable("GroupModel", (table) => {
    table.increments("id");
    table.string("name", 255).notNullable();
    table.string("privileges", 51).notNullable();
    table.datetime("created").notNullable();
    table.index(["privileges"], "");
})
  .createTable("UserModel", (table) => {
    table.increments("id");
    table.string("user", 255).notNullable();
    table.string("password", 51).nullable();
    table.integer("age").nullable();
    table.datetime("created").notNullable();
    table.index(["user"], "user_idx");
});

exports.down = (knex) => knex.schema.dropTable("GroupModel")
  .dropTable("UserModel");
"
`;

exports[`Integration tests of migrations makemigrations New database from 0 1`] = `
{
  "db": {
    "UserModel": {
      "className": "UserModel",
      "columns": [
        {
          "autoincrement": true,
          "name": "id",
          "nullable": false,
          "primaryKey": true,
          "property": "id",
          "type": "int",
        },
        {
          "length": 255,
          "name": "user",
          "nullable": false,
          "property": "user",
          "type": "varchar",
        },
        {
          "length": 51,
          "name": "password",
          "nullable": true,
          "property": "password",
          "type": "char",
        },
        {
          "name": "age",
          "nullable": true,
          "property": "age",
          "type": "int",
        },
        {
          "name": "created",
          "nullable": false,
          "property": "created",
          "type": "datetime",
        },
      ],
      "indexes": [
        {
          "name": "user_idx",
          "properties": [
            "user",
          ],
        },
      ],
      "name": "UserModel",
    },
  },
  "knestVersion": "",
  "version": 1,
}
`;

exports[`Integration tests of migrations makemigrations New database from 0 2`] = `
"
exports.up = (knex) => knex.schema.createTable("UserModel", (table) => {
    table.increments("id");
    table.string("user", 255).notNullable();
    table.string("password", 51).nullable();
    table.integer("age").nullable();
    table.datetime("created").notNullable();
    table.index(["user"], "user_idx");
});

exports.down = (knex) => knex.schema.dropTable("UserModel");
"
`;
